core/templates/domain/user/user-profile-object.factory.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Frontend domain object factory for users.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import {
+  LearnerExplorationSummary,
+  LearnerExplorationSummaryBackendDict,
+  LearnerExplorationSummaryObjectFactory
+} from 'domain/summary/learner-exploration-summary-object.factory';
+
+export interface UserProfileBackendDict {
+  'username': string;
+  'profile_is_of_current_user': boolean;
+  'profile_username': string;
+  'user_bio': string;
+  'subject_interests': string[];
+  'first_contribution_msec': number;
+  'profile_picture_data_url': string;
+  'user_impact_score': number;
+  'is_already_subscribed': boolean;
+  'is_user_visiting_own_profile': boolean;
+  'created_exp_summary_dicts': LearnerExplorationSummaryBackendDict[];
+  'edited_exp_summary_dicts': LearnerExplorationSummaryBackendDict[];
+}
+
+export class UserProfile {
+  constructor(
+    public currentUser: string,
+    public profileIsOfCurrentUser: boolean,
+    public profileUsername: string,
+    public userBio: string,
+    public subjectInterests: string[],
+    public firstContributionMsec: number,
+    public profilePictureDataUrl: string,
+    public userImpactScore: number,
+    public isAlreadySubscribed: boolean,
+    public isUserVisitingOwnProfile: boolean,
+    public createdExpSummaries: LearnerExplorationSummary[],
+    public editedExpSummaries: LearnerExplorationSummary[]
+  ) {}
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UserProfileObjectFactory {
+  constructor(
+    private learnerExplorationSummaryObjectFactory:
+    LearnerExplorationSummaryObjectFactory) {}
+
+  createFromBackendDict(backendDict: UserProfileBackendDict) {
+    return new UserProfile(
+      backendDict.username,
```

Comment:
Woah. Why do we have "username" and "profile username"? That's just confusing.

Are both needed? What is the difference between the two fields?

Also you should not be renaming this to "currentUser". Please be really careful about the names of things. A "user" and a "username" are different. I've seen this pattern in a couple places now so I'd like to request that you audit all the fields and feel free to ask questions if you are not sure whether a renaming is valid.

Reply:
profile usename -> the username of profile being visited.
username -> This is the username of the user currently logged in. This is sent along with almost every endpoint.

> Are both needed

This is used in that page just to determine if any user is logged in.

Should I change it to `currentUsername` instead?

Reply:
username and usernameOfViewedProfile should work, I think. Best to keep "username" consistent everywhere.

Also, you have not replied to my last paragraph. Please make sure to respond to **all** parts in a comment. I would like a confirmation that you have reviewed every field -- probably what makes the most sense is for you to please list all the name changes explicitly here, with links to the code for each of those.

Reply:
> username and usernameOfViewedProfile should work, I think. Best to keep "username" consistent everywhere.

done.

Ah, sorry! I think I missed the last para. I had some doubts with it.

> I've seen this pattern in a couple places now so I'd like to request that you audit all the fields and feel free to ask questions if you are not sure whether a renaming is valid.

I didn't actually understand which pattern?

Reply:
The pattern of renaming fields to different names (i.e. anywhere where there's any sort of difference between the key name in the backend dict, the attribute name of the frontend object, or the variable used to refer to a value in the dict/object).

Does that clarify?

Reply:
@seanlip i had some questions
1. does that mean that in some cases like this one to change the name of field in backend_dict too? (changing profile_username to username_of_viewed_profile)
2. In some cases the backend_dict says skill_id and in the domain it is changed to id. Should these kind of cases be also changed?

Reply:
Renaming is fine in some cases as long as it is not confusing. Hence the request for a list of all instances so that I can give you specific advice.

Reply:
Here are instances I could find

https://github.com/oppia/oppia/blob/f3ec0b4e15059ab8e6b452abca8bb9e59b6292bf/core/templates/domain/topic_viewer/read-only-topic-object.factory.ts#L38-L61

Here properties like `additional_story_dicts` is renamed to `additionalStorySummaries`. I personally think, this is okay!


This file -> https://github.com/oppia/oppia/blob/develop/core/templates/domain/user/profile-summary-object.factory.ts
This is a common object factory for two backend dicts because it doesn't make sense to make a separate objects for these dicts.

Here -> https://github.com/oppia/oppia/blob/develop/core/templates/domain/exploration/RuleObjectFactory.ts
`rule_type` is renamed to just `type`. 

Here -> https://github.com/oppia/oppia/blob/develop/core/templates/domain/question/QuestionSummaryObjectFactory.ts
I've corrected this one!

Here -> https://github.com/oppia/oppia/blob/develop/core/templates/domain/skill/SkillSummaryObjectFactory.ts
I think thats okay!

Here -> https://github.com/oppia/oppia/blob/develop/core/templates/domain/skill/SkillRightsObjectFactory.ts
Corrected this one!

Here -> https://github.com/oppia/oppia/blob/develop/core/templates/domain/feedback_thread/FeedbackThreadSummaryObjectFactory.ts
Corrected this one!

Reply:
@seanlip ptal!